File: fanboi2/models/board.py
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import synonym
from sqlalchemy.sql import func
from sqlalchemy.sql.schema import Column
from sqlalchemy.sql.sqltypes import DateTime, Integer, String, Text, Unicode, JSON

from ._base import Base, Versioned
from ._type import BoardStatusEnum


DEFAULT_BOARD_CONFIG = {
    "name": "Nameless Fanboi",
    "use_ident": True,
    "max_posts": 1000,
    "post_delay": 10,
    "expire_duration": 0,
}


class Board(Versioned, Base):
    """Model class for board. This model serve as a category to topic and
    also holds settings regarding how posts are created and displayed. It
    should always be accessed using :attr:`slug`.
    """

    __tablename__ = "board"

    id = Column(Integer, primary_key=True)
    created_at = Column(DateTime(timezone=True), default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    slug = Column(String(64), unique=True, nullable=False)
    title = Column(Unicode(255), nullable=False)
    _settings = Column("settings", JSON, nullable=False, default={})
    agreements = Column(Text, nullable=True)
    description = Column(Text, nullable=True)
    status = Column(BoardStatusEnum, default="open", nullable=False)

    def get_settings(self):
        if self._settings is None:
            return DEFAULT_BOARD_CONFIG
        settings = DEFAULT_BOARD_CONFIG.copy()
        settings.update(self._settings)
        return settings

    def set_settings(self, value):
        self._settings = value

    @declared_attr
    def settings(self):
        return synonym(
            "_settings", descriptor=property(self.get_settings, self.set_settings)
        )

File: assets/javascripts/models/board.ts
import { Model, IModelData } from "./base";
import { Topic } from "./topic";
import { CancellableToken } from "../utils/cancellable";
import { request } from "../utils/request";

export class Board extends Model {
    type: string;
    id: number;
    agreements: string;
    description: string;
    settings: {
        postDelay: number;
        useIdent: boolean;
        name: string;
        maxPosts: number;
    };
    slug: string;
    title: string;
    path: string;

    protected serialize(data: IModelData) {
        Model.assertType(data, "board");

        this.type = data["type"];
        this.id = data["id"];
        this.agreements = data["agreements"];
        this.description = data["description"];
        this.slug = data["slug"];
        this.title = data["title"];
        this.path = data["path"];

        this.settings = {
            postDelay: data["settings"]["post_delay"],
            useIdent: data["settings"]["use_ident"],
            name: data["settings"]["name"],
            maxPosts: data["settings"]["max_posts"],
        };
    }

    static queryAll(token?: CancellableToken): Promise<Board[]> {
        return request("GET", "/api/1.0/boards/", {}, token).then(
            (resp: string): Board[] => {
                return JSON.parse(resp).map((data: Object) => {
                    return new Board(data);
                });
            },
        );
    }

    static querySlug(slug: string, token?: CancellableToken): Promise<Board> {
        return request("GET", `/api/1.0/boards/${slug}/`, {}, token).then(
            (resp: string): Board => {
                return new Board(JSON.parse(resp));
            },
        );
    }

    getTopics(token?: CancellableToken): Promise<Topic[]> {
        return Topic.queryAll(this.slug, token);
    }
}
